File: GFG POTD/022-September-2025/012-minimize-the-heights.java
// https://www.geeksforgeeks.org/problems/minimize-the-heights3351/1/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(1)
    public int getMinDiff(int[] arr, int k) {
        int n = arr.length;
        
        Arrays.sort(arr);
        
        int maxEle = arr[n - 1], minEle = arr[0];
        int minDiff = maxEle - minEle;
        
        int maxi = 0, mini = 0;
        
        for (int i = 1; i < n; i ++) {
            if (arr[i] >= k) {
                maxi = Math.max(maxEle - k, arr[i - 1] + k);
                mini = Math.min(minEle + k, arr[i] - k);
                
                minDiff = Math.min(minDiff, maxi - mini);
            }
        }
        
        return minDiff;
    }
}

File: GFG POTD/022-September-2025/012-minimize-the-heights.cpp
// https://www.geeksforgeeks.org/problems/minimize-the-heights3351/1/




#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // TC : O(N)
    // SC : O(1)
    int getMinDiff(vector<int> &arr, int k) {
        int n = arr.size();
        sort(begin(arr), end(arr));
        
        int maxi = arr[n - 1];
        int mini = arr[0];
        int maxDiff = maxi - mini;
        
        int maxEle = 0, minEle = 0;
        
        for (int i = 1; i < n; i ++) {
            if (arr[i] >= k) {
                maxEle  = max(arr[i-1] + k, maxi - k);  // try to reduce maximum
                minEle  = min(arr[i] - k  , mini + k);  // try to increase minimum
                
                maxDiff = min(maxDiff, maxEle - minEle);
            }
        }
        
        return maxDiff;
    }
};
